File: OmokServer/packet_buffer_manager.cpp
#include "packet_buffer_manager.h"

std::mutex PacketBufferManager::mutex_;
char* PacketBufferManager::packet_buffer_;
char* PacketBufferManager::packet_buffer_temp_;
uint32_t PacketBufferManager::buffer_size_;
uint32_t PacketBufferManager::read_pos_;
uint32_t PacketBufferManager::write_pos_;
uint32_t PacketBufferManager::header_size_;
uint32_t PacketBufferManager::max_packet_size_;

void PacketBufferManager::ClearBuffer() {
	delete[] packet_buffer_;
	delete[] packet_buffer_temp_;
};

void PacketBufferManager::Init(uint32_t buffer_size, uint32_t header_size, uint32_t max_packet_size)
{
	packet_buffer_ = new char[buffer_size];
	packet_buffer_temp_ = new char[buffer_size];
	buffer_size_ = buffer_size;
	header_size_ = header_size;
	max_packet_size_ = max_packet_size;
}

bool PacketBufferManager::Write(char* data, uint32_t size)
{
	// 하나의 스레드만 버퍼에 접근하도록 락
	std::lock_guard<std::mutex> lock(mutex_);
	
	auto remain_buffer_size = buffer_size_ - write_pos_;
	if (remain_buffer_size < size) {
		std::print("Buffer is full\n");
		return false;
	}

	std::copy(data, data + size, packet_buffer_ + write_pos_);
	write_pos_ += size;

	if (HasEnoughBufferSpace()  == false) {
		BufferRelocate();
	}
	return true;
}

char* PacketBufferManager::Read() {
	// 하나의 스레드만 버퍼에 접근하도록 락
	std::lock_guard<std::mutex> lock(mutex_);

	auto readable_size = write_pos_ - read_pos_;
	if (readable_size < header_size_) {
		return nullptr;
	}

	char size[2] = { (packet_buffer_ + read_pos_)[0], (packet_buffer_ + read_pos_)[1] };
	auto packet_size = *reinterpret_cast<uint16_t*>(size);

	std::print("pktSize : {}\n", packet_size);

	if (readable_size < packet_size) {
		return nullptr;
	}

	auto packet_data = packet_buffer_ + read_pos_;
	read_pos_ += packet_size;
	return packet_data;
}

bool PacketBufferManager::HasEnoughBufferSpace()
{
	return (buffer_size_ - read_pos_) >= max_packet_size_;
}

void PacketBufferManager::BufferRelocate()
{
	std::copy(packet_buffer_ + read_pos_, packet_buffer_ + write_pos_, packet_buffer_temp_);

	auto readable_size = write_pos_ - read_pos_;
	std::copy(packet_buffer_temp_, packet_buffer_temp_ + readable_size, packet_buffer_);

	read_pos_ = 0;
	write_pos_ = readable_size;
}

File: OmokClient/PacketBufferManager.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace csharp_test_client
{
    class PacketBufferManager
    {
        int BufferSize = 0;
        int ReadPos = 0;
        int WritePos = 0;

        int HeaderSize = 0;
        int MaxPacketSize = 0;
        byte[] PacketData;
        byte[] PacketDataTemp;

        public bool Init(int size, int headerSize, int maxPacketSize)
        {
            if (size < (maxPacketSize * 2) || size < 1 || headerSize < 1 || maxPacketSize < 1)
            {
                return false;
            }

            BufferSize = size;
            PacketData = new byte[size];
            PacketDataTemp = new byte[size];
            HeaderSize = headerSize;
            MaxPacketSize = maxPacketSize;

            return true;
        }

        public void Clear()
        {
            WritePos = ReadPos = 0;
        }

        public bool Write(byte[] data, int pos, int size)
        {
            if (data == null || (data.Length < (pos + size)))
            {
                return false;
            }

            var remainBufferSize = BufferSize - WritePos;

            if (remainBufferSize < size)
            {
                return false;
            }

            Buffer.BlockCopy(data, pos, PacketData, WritePos, size);
            WritePos += size;

            if (NextFree() == false)
            {
                BufferRelocate();
            }
            return true;
        }

        public ArraySegment<byte> Read()
        {
            var enableReadSize = WritePos - ReadPos;

            if (enableReadSize < HeaderSize)// 패킷이 헤더보다도 안들어옴
            {
                return new ArraySegment<byte>();
            }

            var packetDataSize = BitConverter.ToInt16(PacketData, ReadPos);
            if (enableReadSize < packetDataSize) // 아직 패킷 다 안들어옴
            {
                return new ArraySegment<byte>();
            }

            var completePacketData = new ArraySegment<byte>(PacketData, ReadPos, packetDataSize);
            ReadPos += packetDataSize;
            return completePacketData;
        }

        bool NextFree()
        {
            var enableWriteSize = BufferSize - WritePos;

            if (enableWriteSize < MaxPacketSize)
            {
                return false;
            }

            return true;
        }

        void BufferRelocate()
        {
            var enableReadSize = WritePos - ReadPos;

            Buffer.BlockCopy(PacketData, ReadPos, PacketDataTemp, 0, enableReadSize);
            Buffer.BlockCopy(PacketDataTemp, 0, PacketData, 0, enableReadSize);

            ReadPos = 0;
            WritePos = enableReadSize;
        }
    }
}
